File: Day 81/Solution.java
// 557. Reverse Words in a String III
// Given a string s, reverse the order of characters in each word within a sentence while still preserving whitespace and initial word order.
class Solution {
    public String reverseWords(String s) {
        // Split on single spaces (input uses single spaces between words)
        String[] words = s.split(" ");
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < words.length; i++) {
            StringBuilder w = new StringBuilder(words[i]);
            sb.append(w.reverse());
            if (i < words.length - 1) {
                sb.append(" ");
            }
        }
        return sb.toString();
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.reverseWords("Let's take LeetCode contest"));
        // s'teL ekat edoCteeL tsetnoc
        System.out.println(sol.reverseWords("Mr Ding"));
        // rM gniD
    }
}

File: Day 81/Solution.py
# 557. Reverse Words in a String III
# Given a string s, reverse the order of characters in each word within a sentence while still preserving whitespace and initial word order.
class Solution:
    def reverseWords(self, s: str) -> str:
        # s.split(" ") keeps exact spacing pattern (spaces preserved in between words)
        words = s.split(" ")
        for i in range(len(words)):
            words[i] = words[i][::-1]
        return " ".join(words)

if __name__ == "__main__":
    sol = Solution()
    print(sol.reverseWords("Let's take LeetCode contest"))
    # s'teL ekat edoCteeL tsetnoc
    print(sol.reverseWords("Mr Ding"))
    # rM gniD
